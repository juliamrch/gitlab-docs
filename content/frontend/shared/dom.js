/* global $ */

/**
 * Returns outerHeight of element **even if it's hidden**
 *
 * NOTE: Uses jQuery because there is no trivial way to do this in
 * vaniall JS, and it's nice that jQuery has a reliable out-of-the-box
 * solution.
 *
 * @param {Element} el
 */
export const getOuterHeight = (el) => $(el).outerHeight();

/**
 * Find the first child of the given element with the given tag name
 *
 * @param {Element} el
 * @param {String} tagName
 * @returns {Element | null} Returns first child that matches the given tagName (or null if not found)
 */
export const findChildByTagName = (el, tagName) =>
  Array.from(el.childNodes).find((x) => x.tagName === tagName.toUpperCase());
