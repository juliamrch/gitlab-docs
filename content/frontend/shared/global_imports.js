import jquery from 'jquery';
import 'bootstrap';

window.$ = jquery;
