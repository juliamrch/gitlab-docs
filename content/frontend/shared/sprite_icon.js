/**
 * Sprite helper adapted from
 * https://gitlab.com/gitlab-org/gitlab/-/blob/master/app/assets/javascripts/lib/utils/common_utils.js#L472
 */

export const spriteIcon = (icon, className = '') => {
  const spritePath = '/assets/images/icons.svg';
  const classAttribute = className.length > 0 ? `class="${className}"` : '';

  return `<svg ${classAttribute}><use xlink:href="${spritePath}#${icon}" /></svg>`;
};
