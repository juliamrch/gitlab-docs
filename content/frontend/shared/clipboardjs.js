/* global $ */

import { spriteIcon } from './sprite_icon';

const createCopyButton = () => {
  const button = document.createElement('button');
  button.className = 'clip-btn';
  button.title = 'Click to copy';
  button.dataset.selector = 'true';
  button.innerHTML = spriteIcon('copy-to-clipboard', 'gl-icon ml-1 mr-1 s16');
  return button;
};

window.addEventListener('load', () => {
  document.querySelectorAll('pre').forEach((element) => {
    const button = createCopyButton();
    element.appendChild(button);
  });

  $('button').tooltip({
    trigger: 'click',
    placement: 'left',
  });

  function setTooltip(btn, message) {
    $(btn).tooltip('hide').attr('data-original-title', message).tooltip('show');
  }

  function hideTooltip(btn) {
    setTimeout(function hide() {
      $(btn).tooltip('hide');
    }, 1000);
  }

  document.querySelectorAll('.clip-btn').forEach((button) => {
    button.addEventListener('click', (event) => {
      const clickedButton = event.target;

      try {
        navigator.clipboard.writeText(clickedButton.previousElementSibling.innerText);
        setTooltip(clickedButton, 'Copied!');
        hideTooltip(clickedButton);
      } catch {
        setTooltip(clickedButton, 'Failed!');
        hideTooltip(clickedButton);
      }
    });
  });
});
