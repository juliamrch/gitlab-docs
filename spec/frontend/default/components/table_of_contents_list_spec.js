import { mount } from '@vue/test-utils';
import { parseTOC } from '../../../../content/frontend/shared/dom_parse_toc';
import TableOfContentsList from '../../../../content/frontend/default/components/table_of_contents_list.vue';
import { createExampleToc } from '../../shared/toc_helper';

describe('frontend/default/components/table_of_contents_list', () => {
  let wrapper;

  const createComponent = (props = {}) => {
    wrapper = mount(TableOfContentsList, {
      propsData: {
        ...props,
      },
    });
  };

  const findItemsData = () => parseTOC(wrapper.element);

  afterEach(() => {
    wrapper.destroy();
    wrapper = null;
  });

  describe('with items', () => {
    let items;

    beforeEach(() => {
      items = createExampleToc();
      createComponent({ items });
    });

    it('renders all items', () => {
      expect(findItemsData()).toEqual(items);
    });
  });

  describe('with empty items', () => {
    beforeEach(() => {
      createComponent({ items: [] });
    });

    it('shows empty ul', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });
});
