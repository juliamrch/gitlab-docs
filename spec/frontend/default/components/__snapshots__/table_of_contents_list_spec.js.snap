// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`frontend/default/components/table_of_contents_list with empty items shows empty ul 1`] = `
<ul
  class="nav nav-pills flex-column"
/>
`;
